actix-tracing/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+}
+
+impl<S, T> TracingService<S, T>
+where
+    S: Service,
+    T: MakeSpan<S::Request>,
+{
+    pub fn new(inner: S, make_span: T) -> Self {
+        TracingService { inner, make_span }
+    }
+}
+
+impl<S, T> Service for TracingService<S, T>
+where
+    S: Service,
+    T: MakeSpan<S::Request>,
+{
+    type Request = S::Request;
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Either<S::Future, Instrumented<S::Future>>;
+
+    fn poll_ready(&mut self, ctx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(ctx)
+    }
+
+    fn call(&mut self, req: Self::Request) -> Self::Future {
+        let span = self.make_span.span_for(&req);
+        let _enter = span.as_ref().map(|s| s.enter());
+
+        let fut = self.inner.call(req);
+
+        // make a child span to track the future's execution
+        if let Some(span) = span
+            .clone()
+            .map(|span| tracing::span!(parent: &span, tracing::Level::INFO, "future"))
+        {
+            Either::Right(fut.instrument(span))
+        } else {
+            Either::Left(fut)
+        }
+    }
+}
+
+/// A `Transform` implementation that wraps services with a [`TracingService`].
+///
+/// [`TracingService`]: struct.TracingService.html
+pub struct TracingTransform<S, U, T> {
+    make_span: T,
+    _p: PhantomData<fn(S, U)>,
+}
+
+impl<S, U, T> TracingTransform<S, U, T> {
+    pub fn new(make_span: T) -> Self {
+        TracingTransform {
+            make_span,
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<S, U, T> Transform<S> for TracingTransform<S, U, T>
+where
+    S: Service,
+    U: ServiceFactory<
+        Request = S::Request,
+        Response = S::Response,
+        Error = S::Error,
+        Service = S,
+    >,
+    T: MakeSpan<S::Request> + Clone,
+{
+    type Request = S::Request;
+    type Response = S::Response;
+    type Error = S::Error;
+    type InitError = U::InitError;
+    type Transform = TracingService<S, T>;
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ok(TracingService::new(service, self.make_span.clone()))
+    }
+}
+
+/// Wraps the provided service factory with a transform that automatically
+/// enters/exits the given span.
+///
+/// The span to be entered/exited can be provided via an implementation of
+/// the [`MakeSpan`] trait. A blanket implementation of [`MakeSpan`] for
+/// `Fn(&T) -> Option<Span>` allows you to pass a closure for the `make_span`
+/// argument. The closure is passed in a reference to the request being
+/// handled by the service.
+///
+/// For example:
+/// ```rust,ignore
+/// let traced_service = trace(
+///     web_service,
+///     |req: &Request| Some(span!(Level::INFO, "request", req.id))
+/// );
+/// ```
+///
+/// [`MakeSpan`]: trait.MakeSpan.html
+pub fn trace<S, U, T>(
+    service_factory: U,
+    make_span: T,
+) -> impl ServiceFactory<
```

Comment:
it won't be possible to use `+Clone`  with `impl Trait` here

Reply:
Not sure I understand this. I was able to add a `Clone` bound on the `Fn`.

Reply:
Resulted factory wouldn’t have clone even if input service factory would have clone

Reply:
Let’s return `ApplyTransform` type instead of impl ServiceFactory

Reply:
Changed the return type to `ApplyTransform`.

Still don't understand the `Clone` comment. The `Clone` bound is on the closure and not on the factory returned from `trace`.

Reply:
but if you want to make clonable service factory, it would be not possible with impl Trait